Redefine threshold and interval

gonzaponte opened the pull request:
This PR address the inconsistency of the definition of threshold and interval. This affects the value of some parameters in the code.
The work performed here is summarized in these points:
- Replace `>` by `>=` when dealing with thresholds.
- Fix inconsistency between the python and cython implementation of `minmax` (use right-open intervals in both).
- Exhaustively test python and cython `minmax` implementations.
- Cosmetics.

Making all intervals right-open creates, nonetheless, an uncomfortable situation when dealing with integer values. We have in our configuration:
```
some_integer_variable_min = 1
some_integer_variable_max = 2
```
Which means that only `some_integer_variable = 1` is accepted, but one may think it accepts both values looking at the configuration. On the other hand, setting both to 1 wouldn't work. I guess one just has to remember this is equivalent to python's `range` function, but it is not obvious to me. Opinions on this topic are very much appreciated.

jacg committed:
Make City abstract_methods raise NotImplementedError

jacg committed:
Reimplement Counter as simple sublcass of Namespace

jacg committed:
Fix CLI override report

In cases here

+ an option was set on the CLI
+ the same option had a setting in some config file

the configuration display erroneously reported that the option was set
in the file rather than on the CLI.

This is fixed here.

jacg committed:
Typo

jacg committed:
Introduce detection of unexpected parameters

jacg committed:
Replace remaining nprint with print_mod

jacg committed:
Sanitize blank lines

jacg committed:
Remove commented-out code

jacg committed:
Replace remaining nmax & first_event with event_range

jacg committed:
Fix filesin -> files_in

jacg committed:
Eliminate print_empty_events (not *used* anywhere)

jacg committed:
Remove unused s1_string, s2_string and s1_rebin_stride

jacg committed:
Remove obsolete exceptions

jacg committed:
Remove pointless passes

`pass` is not required if there is *any* code in the block at all. A
docstring is code!

abotas committed:
writers mark columns to be indexed instead of indexing them

abotas committed:
create city.index_tables() and call in city.end()

jacg committed:
Replace assert with raise UnknownParameter

jacg committed:
Add test for detection of unknown parameters

jacg committed:
Cosmetics

gonzaponte committed:
Cosmetics

gonzaponte committed:
rename parametrized test

gonzaponte committed:
Add scope to fixture

gonzaponte committed:
Fix double indentation

gonzaponte committed:
Create a dummy city for testing

abotas committed:
Index pytables after populating them

By advice of F. Alted, all tables are indexed after populating
them. This is faster and more natural.

Due to a bug in pytables' flush machinery, the tables must be indexed
after the file is closed and reopened in order to ensure proper
indexation and no data loss. This shall be reviewed after the bug is fixed.

jacg committed:
Detect unknown parameters

Given that we now treat the parameter namespaces completly dynamically
with **kwds; typos, obsolete parameter names and other mistakes pass
silently. This leads to lots of annoying and difficult to track bugs
and confusion.

This PR introduces detection of unexpected parameters:

+ Each city may provide a class attribute named parameters, being a
  sequence of parameter names that are recognized at this level, but
  not recognized by the superclass.

+ At construction time, the construction arguments are checked for the
  presence of parameters which have not been declared at the city's
  level or its superclasses. If an unexpected parameter is found, an
  AssertionError is raised. (This should probably be replaced by a new
  IC exception.)

jjgomezcadenas committed:
Fix counters in Dorothea.

Inconsistent counters in Dorothea have been fixed.

jjgomezcadenas committed:
Propagate chance in counter name to test in Dorothea.

jjgomezcadenas committed:
Add event_loop to PCity.

Event loop is common to all PCities in particular to Dorothea and Penthesilea.

jjgomezcadenas committed:
Eliminate write_parameters from interface.

write_parameters is specific of DIOMIRA and should not be in the interface

jjgomezcadenas committed:
Add two functions to PCity interface.

In order to treat Dorothea and Penthesilea (plus any other city
derived from PCity) uniformly, we need to provide an interface for the
type of DST that the city will create/write (which is different for
each city) and the type of selector (which is configured differently
for each city).

jjgomezcadenas committed:
Define function filter_event().

Function filter_event() encapsulates selection procedure.

jjgomezcadenas committed:
Simplify Dorothea.

Dorothea code just implements a few interfaces. One sep closer to a
decent code.

jjgomezcadenas committed:
Add compulsory methods to DummyCity.

Methods get_writers and write_parameters must be implemented by all
cities including the dummies.

jjgomezcadenas committed:
Add default write_parameters function to Isidora.

jjgomezcadenas committed:
Refactor (simplify) Dorothea and Penthesilea.

Dotty and Patty loose a lot of fat!

jjgomezcadenas committed:
Refactor Pcity to express common functionality.

Dorothea and Penthesilea are both derived from PCity, since they read
PMAPS. They do many things in common, including filtering. PCity
allows now to express that common functionality.

jjgomezcadenas committed:
Refactor Penthesilea and Dorothea

This PR simplifies and improves a few of our cities and base cities. Replaces the IC exceptions with NotImplementedErrore. Dorothea and Penthesilea have a lot in common and their event loops showed it. This PR squashes redundancies into a more versatile PCity. Penthesilea and Dorothea now need only worry themselves with setting writers and passing them their specific dst_event which are then used by PCity's event loop

jacg committed:
Sanitize bizarre setting of daemons attribute

+ Instead of being done in the constructor, where it belongs, it was
  being done from outside, in `drive`.

+ A lack of daemons should be indicated by the daemons list being
  empty. Instead, it was being treated as a special case, by not
  setting the daemons attribute at all.

+ Instead of using `dict.get()` with a sensible default, the lack of a
  daemons config value was being handled in a weird ad-hoc fashion.

Bonus: the name `d_list_name` was semantically misleading.

jacg committed:
Remove now-superfluous hasattr before daemons loop

Now that a lack of daemons is indicated by an empty list of daemons,
rather than a lack of daemons attribute on the city, the loop does not
need to be protected by `hasattr('daemons')` any more: looping over
the empty list does nothing.

jacg committed:
Turn daemons list into a tuple

It should never be modified after initial setting, so a tuple is more
appropriate.

jacg committed:
Fix misleading function name invoke_daemon -> summon_daemon

jacg committed:
Stop suppressing ModuleNotFoundError for unknown daemons

jacg committed:
Remove commented-out code

jacg committed:
Add missing whitespace in type annotations

jacg committed:
Remove pointless (and sometimes incorrect!) docstrings

jacg committed:
Cosmetics

jacg committed:
Indent incorrectly positioned method

jacg committed:
Remove unused and duplicated make_tracks function

1. It's just a copy'n' paste of something that exists in `paolina_functions`
2. No tests fail when it is removed.

jacg committed:
Fix incorrect return type annotation in compute_blobs

jacg committed:
Remove pointless compexity in voxels_from_track_graph

`track.nodes()` returns a list of voxels in the track graph.

The previous version of this function

1. Looped over this list,
2. unpacked the voxels into their components
3. created new voxels out of the components
4. stuck the new voxels into a new list.

This new list is effectively a copy of the original list.

So why not simply return the original?

No tests fail when this change is made.

abotas committed:
penthesilea test

jmbenlloch committed:
Fix dependencies versions

gonzaponte committed:
Add machinery to perform lifetime corrections depending on XY

gonzaponte committed:
Add S12SelectorOutput class to hold full output from filter

gonzaponte committed:
Make filters a package

gonzaponte committed:
Return full output in S12Selector

gonzaponte committed:
Add s1s2si_filter

It combines s1s2_filter and s2si_filter

gonzaponte committed:
Use full output in base class

gonzaponte committed:
Propagate full output to concrete cities

gonzaponte committed:
Add tests for S12SelectorOutput

gonzaponte committed:
Add tests for filters

gonzaponte committed:
Add tests for S12Selector

gonzaponte committed:
Return full output in s1s2_filter and s2si_filter

gonzaponte committed:
Cosmetics and cleanup

gonzaponte committed:
Add test for Dorothea

gonzaponte committed:
Add test for Penthesilea

gonzaponte committed:
Add file for tests

gonzaponte committed:
Update test file

gonzaponte committed:
Add function to load lifetime xy corrections

gonzaponte committed:
Add test for load_lifetime_xy_corrections

gonzaponte committed:
Add docstring and type hints

jmbenlloch committed:
Update DB to use new tables for SiPM noise

jmbenlloch committed:
Fix conda version to one known to work

Anaconda introduced a bug in version
4.3.27 (https://stackoverflow.com/questions/46450912/unable-to-execute-x86-64-conda-cos6-linux-gnu-gcc-no-such-file-or-directory)
and in general we should keep our software versions under control,
this is a first step towards that.

gonzaponte committed:
Generalize xy_writer and define concrete usages

jmbenlloch committed:
Include test for SipmNoisePDF in DB

This test creates a sample DB with unsorted rows to make sure the data
is read in the correct order. This is done by first reading the values
of the noise bins and then reading the probabilities for all sensors
in the same order as the noise bins.

gonzaponte committed:
s1s2selector full output

So far, the rejected peaks of an event passing the S12Selector were
being propagated to the rest of the code, without knowing which ones
they were. This had two side effects:

We always picked the first S1, but in the case of multiple S1s, we
didn't have the certainty that it would be the right one.

When multiple S2s are present, all of them were being written to file,
independently of whether they were selected or not. Thus, on top of
the extra disk usage, we were having a wrong number of S2s associated
to the event.

This PR incorporates the result of the selection over each peak in the
event to the output of the S12Selector. Other functions and the cities
involved are now aware of it and use the complete output to ignore the
rejected peaks.

Furthermore, a bunch of tests have been added to the S12Selector
class, which was untested as well as the functions defined to use
it. The test suite for the cities Dorothea and Penthesilea have been
also extended with some new tests to ensure the selection is performed
correctly.

jahernando committed:
Add a test for issue 347

jahernando committed:
Init a counter in KrCity to solve issue 347

jmbenlloch committed:
Update SipmNoisePDF test

jmbenlloch committed:
Improve database structure for sipm noise

paolafer committed:
Save hits of all particles

jacg committed:
Use BHit rather than Voxel in bunch_of_hits

jacg committed:
PEP8ify function spacing in paolina_functions_test.py

jacg committed:
Make __str__ of BHit and subclasses more informative

jacg committed:
Add test_voxelize_hits_respects_voxel_dimensions

jacg committed:
Fix voxelize_hits

The original algorithm calculated how many voxels were needed to cover
the whole bounding box of hits. This number was used to specify the
number of bins (voxels) into which the hits should be grouped, without
specifying the range of the bins. So `np.histogramdd` picked the range
of hit positions as the range of the bins. As this is very unlikely to
be an exact multiple of the desired voxel size, the voxels tended to
come out a bit smaller than required.

The solution is to explicity specify the spatial range which the bins
should cover, and ensure that it covers an exact multiple of voxel
sizes.

gonzaponte committed:
Fix docstring typo

gonzaponte committed:
Remove useless instruction

gonzaponte committed:
Make the output of ComputeThresholds a bit more readable

gonzaponte committed:
Accept both a common threshold and one for each sipm in signal_sipm

gonzaponte committed:
Add a flag in CalibratedCity to decide between two threshold modes

The threshold for the SiPMs can be either global, i.e. a common
value for all the sensors or relative to the sensor's noise PDF.

gonzaponte committed:
Set flag in config file

gonzaponte committed:
XY-dependent lifetime corrections

This PR adds an interface to perform XY-dependent lifetime corrections.

gonzaponte committed:
Parametrize a previous test to use both types of thresholds

gonzaponte committed:
Use >= rather than > when finding the inverse of the cdf

gonzaponte committed:
Add tests for signal_sipm

gonzaponte committed:
Pep8-ify spaces between functions

gonzaponte committed:
Allow full configuration when loading xy corrections

gonzaponte committed:
Generalize the normalization options parameter

gonzaponte committed:
Include a new normalization to a fixed number

gonzaponte committed:
Adapt the rest of the code to generalized options

gonzaponte committed:
Add test for the new normalization option

gonzaponte committed:
Raise exceptions when an incomplete set of arguments is given

gonzaponte committed:
Add test to verify exceptions are raised

gonzaponte committed:
Store corrections without normalization

gonzaponte committed:
Add test to ensure safety checks are used

gonzaponte committed:
Make intervals half-open

This corresponds to a general definition of threshold and bin.

Threshold means that it is the minimum value accepted,
therefore >= is the right operator, particularly when dealing
with discrete values.

In the case of bins, we need a definition that does not
produce any misinterpretation. If we choose open intervals
and we define two ranges (a,b) and (b,c), the value b doesn't
fall in any of those, which doesn't feel right. On the other
hand if we choose close intervals and take [a,b] and [b,c],
there is ambiguity for the value b, which falls in both
cathegories. The right choice is a half-open interval. There
are, of course, two options, depending on which side you
want to be open. Following the reasoning for the threshold,
it seems natural to make the interval right-open rather than
left-open.

Besides, this fixes a mismatch between the python and the cython
implementations of the class minmax. This would have been
caught by the tests, provided they'd been there.

paolafer committed:
Add test to retrieve all hits

gonzaponte committed:
Fix tests after redefinition of the intervals

gonzaponte committed:
Fix wrong name in test

jahernando committed:
add assert in dorothea_test.py to solve issue 348

gonzaponte committed:
Cosmetics

gonzaponte committed:
Fix configs

gonzaponte committed:
Cleanup of config files

jmbenlloch committed:
Test to check mc_track_writer works for non-consecutive events

jmbenlloch committed:
Fix mc_track_writer to work with non-consecutive events

gonzaponte committed:
Add tests for minmax contains method

gonzaponte committed:
Add tests for all remaining methods of class minmax

gonzaponte committed:
Make variable a bit more readable

gonzaponte committed:
Remove useless import

gonzaponte committed:
Remove duplicated tests with a combined strategy

jmbenlloch committed:
Fix mc_track_writer to work with non-consecutive events

jahernando committed:
Fix uninitialized counter in Dorothea

paolafer committed:
Save hits of all particles

The compute_mchits_dict() function was selecting only hits coming from
electrons, while the correct behaviour is that all hits are
returned. This PR fixes this bug.

gonzaponte committed:
Avoid reinvention of the wheel

gonzaponte committed:
Simplify good_waveform property

gonzaponte committed:
Replace common code with a function

gonzaponte committed:
Remove useless if

gonzaponte committed:
Use signal_above_threshold function to simplify other methods

gonzaponte committed:
Make the str representation of Peak more readable

gonzaponte committed:
Remove commented out lines

gonzaponte committed:
Gather all try/except structures in class methods

The code is cleaner this way. A lot of try/except structures
have been removed from class methods because they now delegate
the searching operations to other methods.

gonzaponte committed:
Make the output of S12 more readable

gonzaponte committed:
Cosmetics

gonzaponte committed:
Combine str method of S1 and S2

gonzaponte committed:
Remove misleading checks

gonzaponte committed:
Remove useless if

gonzaponte committed:
Add compile_and_test_par function

It was mentioned in the help and called when
running bash manage.sh compile_and_test_par
but not implemented.

gonzaponte committed:
Complete manage.sh help

jacg committed:
Clean up test_voxels_from_track_return_node_voxels

The purpose of the function is to act as a dead ringer for
`graph.nodes()`, providing a name which has a higher-level meaning in
IC.

The original test invoked a number of orthogonal IC functions. The new
version generates random graphs with hypothesis and checks that both
`test_voxels_from_track_return_node_voxels(graph)` and `graph.nodes()`
give identical output.

gonzaponte committed:
Update xy_corrections

- Adds the normalization by a constant to the class `Correction`.
- Generalizes the class so this new normalization fits better.
- Adds safety checks to the class.
- Provides tests for all of the above.
- Changes the philosophy of correction factors normalization.
The normalization now occurs when loading from file.

gonzaponte committed:
Document new parameter

gonzaponte committed:
Export inner methods of the class NoiseSampler into functions

This makes testing much easier.

gonzaponte committed:
Add tests for new functions

jacg committed:
Clean up miscellaneous issues

This PR contains a miscellaneous collection of fixes and cleanups that
emerged out of my post-hoc review of a number of PRs that were
accepted over the summer. This has been sitting in my repo over the
last weeks as I have been otherwise occupied. As I don't expect to
have time in the near future to pursue these any further, I submit it
as it is.

gonzaponte committed:
Add test to ensure correct sample shape

gonzaponte committed:
Create fixtures for time-consuming operations

gonzaponte committed:
Parametrize fixture to work with two modes of the sampler

gonzaponte committed:
Fix noise sampling for masked sensors

gonzaponte committed:
Cosmetics

gonzaponte committed:
Define inner method for readability

gonzaponte committed:
Make the run number a fixture

gonzaponte committed:
Change run number

gonzaponte committed:
Add test to ensure noise probabilities are loaded correctly

gonzaponte committed:
Add sampling test

gonzaponte committed:
Add test to compute_thresholds

gonzaponte committed:
Refactor NoiseSampler class using external functions

gonzaponte committed:
Parametrize previous test

gonzaponte committed:
Cosmetics

gonzaponte committed:
Move true information to fixture

gonzaponte committed:
Fix test that was failing way too often

jacg commented on invisible_cities/types/ic_types_test.py:
```diff
@@ -1,55 +1,69 @@
+    return minmax_c(a, b)
+
+
+def make_xy(a,b):
```
Why do we want `make_xy` and `make_xy_c` as synonyms for `xy` and `xy_c`?

gonzaponte replied:
We don't. I didn't realize it was purely an alias.

jacg replied:
Well, strictly speaking, it's *not* an alias, because `make_xy` is a function, while `xy` is a type. Consequently, even though

    make_xy(a,b) == xy(a,b)

and

    type(make_xy(a,b)) == type(xy(a,b))

unfortunately

    type(make_xy(1,2)) != make_xy

whereas, quite sensibly

    type(xy(1,2)) == xy

gonzaponte replied:
I am not very good with the nomenclature, but you get the point. The correct word would be... a wrapper?

jacg replied:
My point isn't about nomenclature. Yes, it's a wrapper, and it's (almost) a synonym, and (almost) an alias. But that's *not* the point.

The point is that it's utterly useless. It only wastes a couple of lines of code, and breaks some functionality.

I'm guessing that you were motivated to write it because you wanted to pass a callable to hypothesis' `builds`. The `xy` types themselves (like all types) are already perfectly good callables themselves (otherwise you wouldn't be able to call them inside your `make_xy` functions!). Like everything else in Python, they are first-class objects (just data), so you can pass them or store them or do whatever you want with them, without the need to wrap them.

In general, when you find yourself writing something of the form 

    def wrapper(*args, **kwds):
        return implementation(*args, **kwds)

or

    lambda *args, **kwds : implementation(*args, **kwds)

then **don't bother** and just use `implementation` itself, directly.

In both the above samples you are creating a function which takes some inputs, passes them on (without looking at them or modifying them at all) to `implementation` and then returns (without looking at or modifying it) whatever `implementation` returns. It matters not one bit whether `implementation` is a type or not.

That's a pointless waste of lines of code, CPU cycles, key presses, developer brain cycles and reader brain cycles.

gonzaponte replied:
I agree 100%, but I didn't write that. I simply didn't realize that function was *just* a wrapper and used it as it was. I will remove it.

jacg replied:
You mean you didn't write the functions `make_xy` and `make_xy_c`?

gonzaponte replied:
Exactly. I did rename the latter, though.

jacg replied:
Ah, OK, at first glance it looked like they appeared in your last commit, but looking more closely I see that they were there already before.

Looks like @jjgomezcadenas made them in mid-June.

@jjgomezcadenas Please do not write pointless wrappers around types.

jacg commented on invisible_cities/types/ic_types_test.py:
```diff
@@ -47,6 +47,31 @@ def test_minmax_does_not_accept_min_greater_than_max(a,b):
+
+@given(minmaxes)
+def test_minmax_eq(mm):
+    assert mm == mm
```
Eh? `assert mm == mm`? You are testing that any minmax equals itself. OK, but that's a far cry from actually testing the correct behaviour of `==`!

gonzaponte replied:
Yes, that's quite embarrassing... Actual test coming up.

jmbenlloch committed:
Stop using dev version of the DB in travis (closes #367)

jacg committed:
Temporary workaround for #357: use older hypothesis

jacg committed:
PR #362: Fixes to manage.sh

This PR:

    Implements a function that was missing
    Adds a command already available to help string

gonzaponte committed:
Remove pointless wrappers

gonzaponte committed:
Recode test_minmax_eq to do something actually interesting

jacg committed:
Swap arg order to make pytest report more meaningful

jacg committed:
Hack around problem of hits on voxels-boundary being lost

Hypothesis found a counterexample for
`test_voxelize_hits_does_not_lose_energy`, in which hits fell
an (almost) exact multiple of voxel sizes apart. Float arithmetic
imprecision then lost some hits, by thinking they lie on the wrong
side of the boundary.

This approach here is to expand the voxel bounding box a bit, in order
to ensure that the hits fall inside the boundary.

I do not like this approach, but it is a cheap solution, and the best
I can do in the available time.

gonzaponte committed:
PR #370 Reduce maximum length of test array

Speed up a test which does not really require long arrays.

https://github.com/nextic/IC/pull/370

approved by jacg

jacg committed:
Add hypothesis profiles with `use_coverage = False`

In some versions of hypothesis, the coverage feature became painfully
slow.

I can't find a way to override the setting on the command line without
creating a new profile.

Adding these profiles allows us to conveniently compare the timeings
of the tests with and without coverage.

jacg committed:
Bump hypothesis version to 3.32.0

The significant performance degradation that appeared in 3.29 has
mostly been regained in 3.32.

jjgomezcadenas committed:
PR #372 Bump hypothesis to 3.32

[jacg]

The serious performance degradation that invaded Travis with the
advent of hypothesis 3.29, has been mostly fixed in hypothesis
3.32. On the other hand, 3.32 makes test_in_range_right_shape run an
order of magnitude faster.

This PR contains a commit, which advances our hypothesis version to
3.32.0

Additionally, two new hypothesis profiles are added: versions of your
existing dev and hard which switch off use_coverage. I think that we
want to keep use_coverage on, but these profiles will make it easier
to investigate performance degradations in the future.

[approved by jjgomezcadenas]

As the previous discussion reveals, this PR appears to solve a part of
our problems.

gonzaponte committed:
Remove signal_above_threshold method

As discussed in PR #366. It is not very prone to
errors and affects the readability of the code.

gonzaponte committed:
Rename variable so it is more meaningful

jjgomezcadenas committed:
PR #366 Refactor pmaps

https://github.com/nextic/IC/pull/366

[gonzaponte]

This is mainly a cleanup and refactoring of the pmaps module, but it
also contains a couple of changes to the external user:

+ So far, functions searching for some SiPM raised `SipmEmptyList`
  when the `Peak` contained no SiPMs, but `SipmNotFound` when there
  are some SiPMs, but not the one we are looking for. It seems the
  `SipmEmptyList` exception doesn't belong there because the function
  doesn't care about how many SiPMs are there, but whether the one we
  asked for is there or not.

+ Some methods that didn't handle exceptions now do, because the
  handling of exceptions has been contained in dedicated methods.

Closes #360.

[approved by jjgomezcadenas]

Since this reviewer was terrified by the new policy of setting the
reviewer's head on the stake, he has tried hard, and commented the
commits one by one. Overall, this PR improves the code readability and
in the case of the grouping of exception I think it also reduces
notably technical debt. This reviewer has found a couple of minor
points. One of them has already been solved by the master in favor of
the defendant. The second point is minor. This reviewer feels that he
can keep his head on his shoulder approving this PR.

gonzaponte committed:
Allow group and node to be passed as parameter

gonzaponte committed:
Add tests to demonstrate kwargs are not passed internally

One of the tests uses a nasty trick so data can be reused.

gonzaponte committed:
Accept and pass keyword arguments

gonzaponte committed:
Parametrize tests via fixture

gonzaponte committed:
Organize cities' parameters

They were enumerated in a derived city rather than in the
base city. This is a problem bacause the base city
uses those parameters in the methods it defines.

gonzaponte committed:
Return a namespace in get_writers method

gonzaponte committed:
Replace test file with one containing MC tracks

gonzaponte committed:
Add true output of dorothea under certain conditions

gonzaponte committed:
Add fixture with the conditions and true output of a dorothea run

gonzaponte committed:
Fix test_dorothea_KrMC

The test has been adapted to the new data, but it also
includes a reformulation of the test in terms of the
true output given by the fixture.

gonzaponte committed:
Fix pmaps tests

They contained hard-coded values based on the previous
test file. Besides fixing the test itself, this commit
removes that dependence.

gonzaponte committed:
Combine common tests and remove hard-coded values

This commit:
- Removes hard-coded values from tests
- Rewrites some tests so they don't depend
  on those values or
- Moves that dependence to the fixture
- Combine tests common to df_to_s1_dict,
  df_to_s2_dict and df_to_s2si_dict
- Renames those tests

gonzaponte committed:
PEP8-ify spacing between functions

gonzaponte committed:
Rewrite dst_io_test in terms of the new fixture

gonzaponte committed:
Rewrite kdst_io_test in terms of the new fixture

gonzaponte committed:
Remove unused fixture

gonzaponte committed:
Remove glorious test

gonzaponte committed:
Remove unused imports

abotas committed:
PR #375 Increase the flexibility of correction functions

Include the ability to choose group and node from which to load data
in load_xy_corrections and load_lifetime_xy_corrections

Accept keyword arguments in wrappers and loading functions. These are
used as the configuration parameters of the class that is built
internally.

Add test for the new functionalities.

gonzaponte committed:
Replace input file in config

gonzaponte committed:
Fix test_config_drive_penthesilea_counters

gonzaponte committed:
Add true output of penthesilea under certain conditions

gonzaponte committed:
Add fixture with the conditions and true output of a penthesilea run

gonzaponte committed:
Reformulate and rename test_penthesilea_run_on_Kr_MC_ipmt_pmaps_5evt

It now uses the new fixture with the true ouput of penthesilea
for a particular configuration. It has been renamed because
there were no ipmt involved.

gonzaponte committed:
Replace star imports with a explicit one

gonzaponte committed:
Update KrMC_pmaps fixture

gonzaponte committed:
Fix test_dorothea_issue_347

The test was being run with run number 0 (the default),
which is interpreted as a MC run. Therefore, when asked
for MC data, a tables.exceptions.NoSuchNodeError is raised.

gonzaponte committed:
Add test to ensure MC tracks are produced when required

gonzaponte committed:
Add fixture to load mc hits from KrMC test files

gonzaponte committed:
PR #379 Fix penthesilea and dorothea inefficiencies

Besides, this commit:
- Changes the name of test_data file to indicate the correct # of events
- Deletes old commented code

gonzaponte committed:
Fix single-line imports

gonzaponte committed:
Remove commented-out line

gonzaponte committed:
Return namedtuples in some fixtures

gonzaponte committed:
Use namedtuples in tests

gonzaponte committed:
Reformulate test

The test
timebin_size_must_be_equal_to_stride_times_25_ns_shows_bug_in_old_data
has been reformulated, as it was unreadable.
It also had a hardcoded evt number that, incredibly,
matched one in the input file. It doesn't depend on that, now.

gonzaponte committed:
Perform test_load_dst properly

Use the function to be tested rather than
take advantage of the fixture.

abotas committed:
PR #376 Replace KrMC_pmaps file

This PR replaces the previous KrMC_pmaps file with a new one. We need it to contain MC tracks (#374) and we've (I've) lost track of the original dataset it has been taken from. Besides the replacement:

a file containing the output of dorothea (with a particular configuration) has been added
a file containing the output of penthesilea (with a particular configuration) has been added
the tests have been adapted to this new file and improved a bit in some cases
the hard-coded dependencies of some tests have been fixed
the fixture that manages the relevant file info has been extended to contain more data
some "duplicated" tests have been combined
On top of that, I have cleaned up a bit the modules I've touched.

abotas committed:
PR #354 Percentual SiPM Threshold

This PR introduces the possibility of using two types of thresholds for the SiPMs:

Common: a single value in pes for all the SiPMs.
Individual: each SiPM has its own threshold based on a common percentage of noise reduction. Since each SiPM has a different PDF, they also have different thresholds.
The code to compute the thresholds based on the percentage was already in IC, but it has been cleaned up a bit. The cython function to extract the signal from the sipms has been adapted to take both types of thresholds and a flag has been introduced in the cities to switch between both modes. The tests included in this PR are, perhaps, still not very exhaustive. I have introduced three:

A test for irene to check that can run in both modes (this is actually a parametrization of a previous test to run with the two values of the flag). However, it does not check the output, as this is more complicated, but it is somehow complemented with:
Two tests for the function signal_sipm in peak_functions, one for each threshold mode. These tests ensure the output of the function is the one we expect.
The configuration files from now on must include a new option thr_sipm_type that must be equal to "common" or to "individual" (case insensitive).

gonzaponte committed:
Accept write_mc_tracks flag in PCity

gonzaponte committed:
Update configs with new flag

gonzaponte committed:
Extend PmapVectors with mc tracks

gonzaponte committed:
Read and pass mc_tracks to event_loop

gonzaponte committed:
Cosmetics

gonzaponte committed:
Fix hdst file structure

The table MCTracks was stored in / rather than in /MC/

gonzaponte committed:
Add test to ensure the mc tracks written by penthesilea are correct

The test is based on the output of the previous one
in order to save some time during the test run.

gonzaponte committed:
Remove unused imports in base_cities.py

gonzaponte committed:
Parametrize test to check both values of the flag

paolafer committed:
Add test for length function

paolafer committed:
Add missing import

paolafer committed:
Make `length` return geometric rather than topological distance

paolafer committed:
Use approx function when comparing almost equal quantities

paolafer committed:
Apply length and find_extrema functions directly to tracks

paolafer committed:
Set irrelevant energies to 1

jacg committed:
Replace assert_almost_equal with pytest.approx

paolafer committed:
PR #352 Respect requested voxel size

https://github.com/nextic/IC/pull/352

+ `voxelize_hits` was not respecting the requested voxel_dimensions.

+ There was no test of this behaviour.

This PR

+ adds a test demonstrating the problem
+ fixes `voxelize_hits` so that it passes the test
+ improves `__str__` in BHit and subclasses

This last point makes it much easer to understand and debug issues such as the main one in this PR.

[approved by paolafer]

Ok, it's all clear! I have used the code and it produces voxels with the requested dimensions. Green light from my side!

jacg committed:
Make `find_extrema` docstring more explicit

jacg committed:
Follow PEP8 on function spacing

jacg committed:
Clarify logic

jacg committed:
Test that `length` does not take shortcuts

One part of this test fails. This is because contiguity is
broken. (The author never tested it, so it never worked, and has only
been causing confusion ever since. This is about to be fixed in an
upcoming PR.)

jacg committed:
PR #382 Calculate geometric rather than topological track length

https://github.com/nextic/IC/pull/382

[paolafer]

I've realized that the `length` function in `paolina_functions`
returned the number of voxels instead of the actual length of the
track. I have modified that function and added a test to make sure the
returned length is the correct one. I have first used the test on the
old code, which failed, and then on the new one, which passes.

[approved by jacg]

This PR introduces two main changes

+ length calculates the geometric rather than the topological distance
  between the extrema of a track.

+ `length` and `find_extrema` accept track graphs, rather than the
  dictionary of distances between.

I'm happy with this, apart from the general brokenness of `contiguity`
but fixing that does not belong in this PR.

jacg committed:
Add tests that prove that contiguity is broken

jacg committed:
Pass the last set of tests but fail test_blobs

jacg committed:
Pass test_blobs

jacg committed:
Remove unused scale parameter in neighbours

jacg committed:
Implement contiguity specification with symbols

jacg committed:
Use new contiguity interface wherever existing tests indicate

jacg committed:
Remove obsolete xfail mark

jacg committed:
Use default contiguity in tests, where appropriate

jacg committed:
Test that length cuts corners to the extent that contiguity permits

jacg committed:
Use symbols directly rather than strings containing their names

jjgomezcadenas committed:
#383 Fix contiguity

https://github.com/nextic/IC/pull/383

The criterion for identifying neigbour voxels used to contain a
hard-wired magic number. Over time this turned into a
user-configurable pramameter, even though the original logic was not
designed to cater for such flexibility: the original only looked for
voxels which share faces; the uncucessfully upgraded version was
supposed to allow accepting those which share either

+ an edge or
+ a face

or those which share either

+ an edge or
+ a face or
+ a corner

This PR fixes the logic and cleans up the interface, by creating a set
of symbols which allow the contiguity to be specified according to its
meaning, rather than by using some numbers.

[approved by jjgomezcadenas]

Brilliantly implemented, totally clarifies an obscure issues that has
crated some trouble. A real lesson in good implementation practice
including the use of symbols rather than strings. Shows very high
levels of Karma.

gonzaponte committed:
Reset last row read in `mc_track_writer`

This is needed in order to the city to work
on multiple input files. The `mc_track_writer`
contains a variable that keeps track of the
last row read. If we move from one file to
another, we need to reset this variable to
zero in order to start reading from the
beginning.

gonzaponte committed:
Temporary fix to failing test.

A PR will be submitted with a proper fix.

jmbenlloch committed:
#380 Write mc tracks in PCity #380

This PR introduces the ability to propagate the MC tracks in the input
file to the output file for cities deriving from PCity. This affects
mainly to Penthesilea, as we haven't had such need in Dorothea.

Moreover, it introduces minor fixes with regard to the organization of
parameters, fixes a small detail in a previous test and applies some
cosmetics to the code. A couple of tests have also been introduced to
ensure this new functionality works properly.

Closes #374
